### DISCUSSION TITLE
Network Map Code Usage and Custom Devices

### DISCUSSION THREAD

**Discussion by VertorixAU** [Koenkk/zigbee2mqtt#5802] [General]
Hi Keonkk

I have used your networkMap.js code as a base to create my own network map of zigbee networks using Node-Red.  Do you want me to reference your source code in the source code of my project?

Also I have created custom zigbee devices.  https://zigbee2mqtt.discourse.group/t/make-your-own-end-devices/1733
I'm going to put some effort into getting them to join zigbee2mqtt and do mandatory Zigbee service discovery.
Example:  
nodeDescReq, activeEPReq and SimpleDescReg.
The plan is to have them automatically join the Zigbee2MQTT as generic type devices.  Temp sensors etc.  Or possibly create custom devices.

**Comment by Koenkk** [comment]
> Do you want me to reference your source code in the source code of my project?

Fine with that.

Looks like a really cool project, I think lots of people are interested in creating their own zigbee devices!

**Reply by VertorixAU** [reply]
Thanks for your quick reply.  

I will spend some spare time on this and let you know when I'm successful.

**Comment by VertorixAU** [comment]
OK so I'm basically stuck where I was last time I attempted this.  I can get the device to successfully join but it fails the interview process nodeDescReq and activeEPReq.  I can edit the database file and add information manually but this defeats the purpose of making a user friendly end device.  I don't really want to implement a full zcl stack such as used in the Z-stack examples (hundreds of lines of code).  I just wanted to have some standard zdp responses for different devices to give the model ID etc. 

What I don't really understand is the zdp requests have to be handled in the application, but I never see the requests (nodeDescReq and activeEPReq) as af_incoming messages to respond too?  So even if I did have the correct frames to send I don't know when to send them.  Do the Aqara devices use a different method of device discovery?  Maybe there is some easier way to send end device descriptions?

**Reply by Koenkk** [reply]
@VertorixAU When sniffing, do you see Zigbee2MQTT sending the node descriptor requests? Responding to them is important, not only to get modelID but also for the type of device (enddevice/router) and power source (battery/AC)

**Comment by VertorixAU** [comment]
I will set up a sniffer and observe the requests.  Do the responses just get put in the payload of a AF_DATA_REQUEST?

**Reply by Koenkk** [reply]
I don't know how to send a response from ZStack itself (only did some development on the coordinator side). For such questions it's better to ask here: https://e2e.ti.com/

**Reply by VertorixAU** [reply]
OK I have sniffed traffic.  The device successfully joins the network and replies to Node Descriptor Request with a Node Descriptor Response.  But the Node Descriptor Request keeps looping and it keeps receiving responses.  I suspect the code is looking for a value, and because it is not present in the Node Descriptor Response it keeps looping and eventually the interview process fails?

![image](https://user-images.githubusercontent.com/51019536/105574832-3981af80-5dbb-11eb-9cb1-75e63ae9225f.png)

**Reply by VertorixAU** [reply]
Here is the Node Descriptor Request and Node Descriptor Response Status = SUCCESS.
Is it the user description or manufacturer code it is looking for?
![image](https://user-images.githubusercontent.com/51019536/105574913-af861680-5dbb-11eb-9935-e32d17a2979b.png)

**Reply by Koenkk** [reply]
Can you provide the herdsman debug log of pairing and the sniff + network key?

To enable herdsman debug logging, see https://www.zigbee2mqtt.io/information/debug.html#zigbee-herdsman-debug-logging

**Comment by VertorixAU** [comment]
OK no worries.  The network key is default Zigbee2MQTT key, the Trust Centre Key is also default.
See attached sniff and Debug Log.
Log is reporting error: "log Interview - node descriptor request failed for '0x00124b001fc2b5bd', attempt 1"

I'm using default ZNP firmware which is meant to support all ZDO.

[Z2M_Debug_Log.txt](https://github.com/Koenkk/zigbee2mqtt/files/5861272/Z2M_Debug_Log.txt)

[sniff.zip](https://github.com/Koenkk/zigbee2mqtt/files/5861273/sniff.zip)

**Comment by VertorixAU** [comment]
Hi Koenkk
Just as an update on this I was able to pair successfully by mimicking another device and sending ZCL: Report Attributes commands the same as a Aqara Lumi Temp sensor.

The device appears in the database file with the correct Model ID for the device I mimicked.  So sending the correct cluster commands seems to have allowed the device to complete the interview process.

I suppose all I have to do now is change the zigbeeModel to another device in the devices.js file and create my own devices.  I have already sent on off cluster commands on a previous device I was testing and it worked.  So it all seems a possibility now.
Thanks

**Comment by VertorixAU** [comment]
Hi Koenkk

I have been able to get a device to join, complete the interview process, and send ZCL report commands (temperature readings).
I believe it was as simple as restarting the ZNP to put it in pairing mode (respond to ZDO Node_desc, Active_EP and Simple Desc and start ZCL interview process).  I also send the  Model Identifier and Application Version using a ZCL: Report Attributes command which may be necessary.
Sniffing the traffic was a key to understanding the interview process.  I didn't realise that it takes a while to complete.  This was a fairly involved process.
Now off to create some custom end devices.
Thanks for your help.

**Comment by VertorixAU** [comment]
I have noticed a lot of Data Requests in the sniffer log for the device I made after a device reset.  I assume this is the end-device polling the coordinator.  I changed the poll rate to 7.5sec and it now does the data requests every 7.5sec after joining.
But after the device resets and joins the Data Requests  occur the same as the rejoin poll rate (440ms).  I don't really get this because the device joins ok with the previous network state restored.  If I erase the network state on reset the device joins ok and polls at the desired rate.  But this would not be idea as it would lose bindings on reset.

Any idea what could be causing this issue?

**Reply by Koenkk** [reply]
I don't have experience with that, please ask this on the Texas Instruments forum: https://e2e.ti.com/

**Reply by VertorixAU** [reply]
Hi Koenkk

I have made 4 different example devices that are compatible with Zigbee2MQTT.  I have chosen 4 different devices which people can use as templates to expand and create their own custom devices.
1. Standard Router:  Sends a periodic attribute report; status online
2. GPIO control device (Router): Allows the user to switch a GPIO output on/off using the OnOff cluster
3. Temperature/Humidity Sensor (End Device): Uses a DHT11 sensor.  The device sleeps with a power consumption around 7uA (extremely low battery drain).  It runs the following loop: 1. take sensor readings, 2. transmit senor data & 3. Sleep
4. Soil Moisture Sensor (End Device):  This device runs the same loop as above.  It uses a capacitive Soil Moisture sensor.  Handy for monitoring indoor plants etc.
All the development boards are programmed in the Arduino IDE and easily customisable.  People can use them as templates to develop their own devices for Zigbee2MQTT.  Can you please send me an e mail at vertorix.au at gmail dot com to discuss how to progress from here.
Thanks
